RecoHGCal/TICL/plugins/TracksterLinkingbyGNN.cc

```diff
@@ -0,0 +1,443 @@
+                    lcs_j.push_back({x, y, z});
+            }
+
+            // Calculate distances
+            auto dists = calculate_dist(lcs_i, lcs_j);
+            float min_dist = *std::min_element(dists.begin(), dists.end());
+            float max_dist = *std::max_element(dists.begin(), dists.end());
+
+            // Calculate delta_r
+            float delta_r = std::sqrt(std::pow((ts_i.barycenter().x() - ts_j.barycenter().x()), 2) +
+                                      std::pow((ts_i.barycenter().y() - ts_j.barycenter().y()), 2));
+
+            // Calculate spatial compatibility
+            float prod = eigenvector0_i.x() * eigenvector0_j.x() +
+                         eigenvector0_i.y() * eigenvector0_j.y() +
+                         eigenvector0_i.z() * eigenvector0_j.z();
+            prod = (prod > 1) ? 1 : prod;
+            prod = (prod < 0) ? 0 : prod;
+            float spatial_comp = std::acos(prod);
+
+            // Calculate time compatibility
+            float time_comp = 0.0;
+            if (ts_i.time() > -99 && ts_j.time() > -99)
+                time_comp = std::abs(ts_i.time() - ts_j.time());
+            else
+                time_comp = -10.0;
+
+            edge_features.push_back(delta_en);      // 0 - energy_difference
+            edge_features.push_back(delta_z);       // 1 - delta_z
+            edge_features.push_back(min_dist);      // 2 - min_dist
+            edge_features.push_back(max_dist);      // 3 - max_dist
+            edge_features.push_back(delta_r);       // 4 - delta_r
+            edge_features.push_back(spatial_comp);  // 5 - spatial_comp
+            edge_features.push_back(time_comp);     // 6 - time_comp
+        }
+    }
+
+    auto numEdges = static_cast<int>(edges_src.size());
+
+    if (numEdges < 1) {
+        LogDebug("HGCalTICL_TracksterLinkingByGNN") << "No edges for the event - no linking is done." << std::endl;
+        // Do not run the network - return the original tracksters
+        linkedTracksterIdToInputTracksterId.resize(N);
+        std::vector<unsigned int> linkedTracksters;
+
+        for (int trackster_id = 0; trackster_id < N; trackster_id++)
+        {
+            linkedTracksterIdToInputTracksterId[trackster_id].push_back(trackster_id);
+            linkedTracksters.push_back(resultTracksters.size());
+            resultTracksters.push_back(input.tracksters[trackster_id]);
+            outputSuperclusters.push_back(linkedTracksters);
+        }
+        return;
+    }
+    
+    // If we use MLP - add additional two featrues
+    if (nnVersion_ == "mlp_edge_features" || nnVersion_ == "mlp_no_edge_features"){
+      auto degrees_centr = countDegree(edges_src, edges_dst, N);
+      for (int i = 0; i < N; i++)
+      {    
+         features.insert(features.begin() + 27 * (i+1), node_degrees[i]/node_degree_max); // 27 - inner node degrees
+         features.insert(features.begin() + 28 * (i+1), degrees_centr[i]); // 28 - degree centrality
```

Comment:
I guess 27 and 28 are the number of features, can we make these numbers as `constexpr` ?